### PR TITLE
jobs,importccl: adopt protected timestamps for IMPORT

### PR DESCRIPTION
**Pull request by ajwerner**
This PR comes in two commits. See each for details.

1) The first refactors and augments the `jobs.Registry` API to enable transactional creation of a job which will be started on the local gateway. I would love to discuss this commit itself a bit.
2) Adopt protected timestamps and demonstrate their effectiveness with a test.

### CODE REVIEW

**Comment by ajwerner** [comment]
cc @lucy-zhang this is the PR with the jobs change

**Review by spaskob — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/jobs/registry.go, line 286 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qcosk5K0LvzIMwbNR:-M-Qcosl2pR7MUp8QBpg:b-iwj2p3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L286)):*
> ```Go
> // StartableJob is a job created with a transaction to be started later.
> // See Registry.CreateStartableJob
> type StartableJob struct {
> ```

please move to jobs.go
___
*[pkg/jobs/registry.go, line 304 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qd5_lDOFLEHKdZ8Z8:-M-Qd5_lDOFLEHKdZ8Z9:bfva47f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L304)):*
> ```Go
> 		}
> 	}()
> 	// TODO(ajwerner): Consider adding some protection against starting more than
> ```

can you add detail on how that can happen?
___
*[pkg/jobs/registry.go, line 320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QdXCM6_o0AUXwPOyO:-M-QdXCM6_o0AUXwPOyP:b-re6loo) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L320)):*
> ```Go
> 
> // CreateStartableJobWithTxn creates a job to be started later, after the
> // creating txn commits. It stores writes job in the jobs table, marks it
> ```

... stores writes ... ?
___
*[pkg/jobs/registry.go, line 332 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QdOdo9CZ1LUx68bhG:-M-QdOdo9CZ1LUx68bhH:b-1adezm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L332)):*
> ```Go
> 		return nil, err
> 	}
> 	j.WithTxn(nil)
> ```

please add a comment why we need this
___
*[pkg/jobs/registry.go, line 338 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qdem3DtuPZSdDP2lB:-M-Qdem3DtuPZSdDP2lC:bpbkhvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L338)):*
> ```Go
> 	}
> 	resumerCtx, cancel := r.makeCtx()
> 	if err := r.register(*j.ID(), cancel); err != nil {
> ```

if the txn is aborted the job will not be unregistered.

**Review by spaskob — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/ccl/importccl/import_stmt.go, line 630 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QeyaO3mmnkGOhnkoY:-M-QeyaO3mmnkGOhnkoZ:b-is6aqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/d965916f753df2673bbc59fbb405872252d94e12/pkg/ccl/importccl/import_stmt.go#L630)):*
> ```
> StartableJob
> ```
why can't we do that here using using QueueJob from planner.go?

**Comment by ajwerner** [comment]
cc @lucy-zhang this has the jobs change we talked about - @spaskob and I discussed offline that you should certainly merge your change that utilized `QueueJob` and `Run` as written today and then later that can be refactored. This doesn't touch that API at all, just an FYI and some reassurance that any API changes in jobs will be out of your way.

File: pkg/ccl/importccl/import_stmt.go [hunk 1]
```diff
@@ -619,20 +623,54 @@ func importPlanHook(
...
+				ProtectedTimestampRecord: protectedTimestampID,
 			},
 			Progress: jobspb.ImportProgress{},
-		})
+		}
+		var sj *jobs.StartableJob
+		// Prepare the protected timestamp record.
+		spans := make([]roachpb.Span, 0, len(tableDetails))
+		for i := range tableDetails {
+			spans = append(spans, tableDetails[i].Desc.TableSpan())
```
Comment (dt) [inline pkg/ccl/importccl/import_stmt.go]: I think we only need to add spans for tables where `IsNew` is false (which could require revert) and then below we can skip if spans is empty

**Review by dt — APPROVED** [review]

**Review by ajwerner — COMMENTED** [review]
This is RFAL. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-:-M12BXzi2kiYWugWo11p:b-n4viu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)

___
*[pkg/ccl/importccl/import_stmt.go, line 630 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QeyaO3mmnkGOhnkoY:-M12AqCv2mMBpsg57sjJ:b8dk3ek) ([raw file](https://github.com/cockroachdb/cockroach/blob/d965916f753df2673bbc59fbb405872252d94e12/pkg/ccl/importccl/import_stmt.go#L630)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

> ```
> StartableJob
> ```
why can't we do that here using using QueueJob from planner.go?
</blockquote></details>

I think we discussed this offline. 
___
*[pkg/ccl/importccl/import_stmt.go, line 652 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-I09eJqLXuJIJd_pS7-r4-652:-M12Anp63gqAczf_9Mtf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d965916f753df2673bbc59fbb405872252d94e12/pkg/ccl/importccl/import_stmt.go#L652)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I think we only need to add spans for tables where `IsNew` is false (which could require revert) and then below we can skip if spans is empty
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 286 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qcosk5K0LvzIMwbNR:-M129_4sCfeOJviL3gl1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L286)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

please move to jobs.go
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 304 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qd5_lDOFLEHKdZ8Z8:-M-SUGK52IazyojqdbSL:b951d6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L304)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

can you add detail on how that can happen?
</blockquote></details>

If there's a programming error in the client. I added protection. 
___
*[pkg/jobs/registry.go, line 320 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QdXCM6_o0AUXwPOyO:-M11m9q79rpUJ6uAVRXg:b-lx4fj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L320)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

... stores writes ... ?
</blockquote></details>

Fixed. 
___
*[pkg/jobs/registry.go, line 332 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-QdOdo9CZ1LUx68bhG:-M129a8f2dn8RfWhB_-D:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L332)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

please add a comment why we need this
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 338 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M-Qdem3DtuPZSdDP2lB:-M11md-L59kqq1EZjehT:b-eomina) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e4aff4f84b985450f052341f382dbce64a664b9/pkg/jobs/registry.go#L338)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

if the txn is aborted the job will not be unregistered.
</blockquote></details>

Good point! I've extended the interface to deal with that case and added some testing. The contract is now that user transactions which are rolled back will need to be cleaned up using `CleanupOnRollback`. I also added unit testing.

**Comment by ajwerner** [comment]
erm actually let me fix the test, must have broken something

**Review by ajwerner — COMMENTED** [review]
Okay, RFAL now

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-:-M12i8Qv8P6By4WjXOj1:ba9yflu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @spaskob)

**Comment by ajwerner** [comment]
@spaskob can you give the first commit this a look? I'd like to get this merged ASAP

**Comment by spaskob** [comment]
ok looking now

**Review by spaskob — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/jobs/jobs.go, line 740 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M16RwfO92e_fNxjdzg6:-M16RwfO92e_fNxjdzg7:b564fy9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0fdbcda070b78940d30b2c300eb3bc9f5848574/pkg/jobs/jobs.go#L740)):*
> ```Go
> }
> 
> // Start will resume a resumable job. The transaction used to create the
> ```

Start will start a job; resume can be confused with `RESUME job_id`
___
*[pkg/jobs/jobs_test.go, line 1994 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M16V1Lp5dLhHnuEI7br:-M16V1Lp5dLhHnuEI7bs:b-ky75yv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0fdbcda070b78940d30b2c300eb3bc9f5848574/pkg/jobs/jobs_test.go#L1994)):*
> ```Go
> }
> 
> // TestStartableJobErrors tests that the StartableJob returns the expected
> ```

I know you want to get this merged but a good to have is to search for use of CreateAndStartJob in this test file sand see if it may be useful to add similar test or replace with a StartableJob.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/jobs/jobs_test.go, line 1994 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M16V1Lp5dLhHnuEI7br:-M16eqwJ-RCtimep0-My:b2fpaqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0fdbcda070b78940d30b2c300eb3bc9f5848574/pkg/jobs/jobs_test.go#L1994)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

I know you want to get this merged but a good to have is to search for use of CreateAndStartJob in this test file sand see if it may be useful to add similar test or replace with a StartableJob.
</blockquote></details>

I don't know that I follow. CreateAndStartJob is implemented on top of StartableJob. All tests which exercise it exercise the StartableJob code.

**Review by spaskob — APPROVED** [review]
ok great then, I missed that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-:-M16fwes8oJytaNCWtwu:b7hrtq4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

**Comment by ajwerner** [comment]
The remaining test failures are addressed by https://github.com/cockroachdb/cockroach/pull/45544

**Review by spaskob — APPROVED** [review]
LGTM for jobs code 

> Reviewed 1 of 7 files at r5, 4 of 9 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-:-M1E9DKr3ssoE3RIxIOM:b-uc3o8p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/ccl/importccl/import_processor_test.go, line 611 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M1E8XQ5DyIhdxpNOOJ2:-M1E8XQ5DyIhdxpNOOJ3:bw459qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d32c56d9f29307eff4c1f8892db55680e96ae0c/pkg/ccl/importccl/import_processor_test.go#L611)):*
> ```
> 
> ```
___
*[pkg/ccl/importccl/import_processor_test.go, line 611 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M1E8YQ-Exawx_c1vG61:-M1E8YQ-Exawx_c1vG62:bw459qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d32c56d9f29307eff4c1f8892db55680e96ae0c/pkg/ccl/importccl/import_processor_test.go#L611)):*
> ```
> 
> ```
___
*[pkg/ccl/importccl/import_processor_test.go, line 612 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M1E8ZqN5HmBkoKwP27R:-M1E8ZqN5HmBkoKwP27S:b3q7y3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d32c56d9f29307eff4c1f8892db55680e96ae0c/pkg/ccl/importccl/import_processor_test.go#L612)):*
> ```
> 			resumer := raw.(*importResumer)
> ```

**Review by ajwerner — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-:-M1H91ZS0sdIBdbUqjjJ:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @spaskob)

___
*[pkg/jobs/jobs.go, line 740 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44739#-M16RwfO92e_fNxjdzg6:-M1H90Of9Td9FiRvyUrE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0fdbcda070b78940d30b2c300eb3bc9f5848574/pkg/jobs/jobs.go#L740)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Start will start a job; resume can be confused with `RESUME job_id`
</blockquote></details>

Done.

**Pull request merged**